operators/src/processing/new_expression/query_processor.rs

```diff
@@ -0,0 +1,401 @@
+                let map_no_data = self.map_no_data;
+
+                let data = tokio::task::spawn_blocking(move || {
+                    thread_pool.install(move || {
+                        Tuple::compute_expression(rasters, &program, map_no_data, out_no_data)
+                    })
+                })
+                .await??;
+
+                let out =
+                    Grid2D::<TO>::new(output_grid_shape, data, Some(self.no_data_value))?.into();
+
+                Ok(RasterTile2D::new(
+                    out_time,
+                    out_tile_position,
+                    out_global_geo_transform,
+                    out,
+                ))
+            });
+
+        Ok(stream.boxed())
+    }
+}
+
+#[async_trait]
+trait ExpressionTupleProcessor<TO: Pixel>: Send + Sync {
+    type Tuple: Send + 'static;
+
+    async fn queries<'a>(
+        &'a self,
+        query: RasterQueryRectangle,
+        ctx: &'a dyn QueryContext,
+    ) -> Result<BoxStream<'a, Result<Self::Tuple>>>;
+
+    fn all_empty(tuple: &Self::Tuple) -> bool;
+
+    fn empty_raster(tuple: &Self::Tuple) -> RasterTile2D<TO>;
+
+    fn metadata(tuple: &Self::Tuple) -> (TimeInterval, GridIdx2D, GeoTransform, GridShape2D);
+
+    fn compute_expression(
+        tuple: Self::Tuple,
+        program: &LinkedExpression,
+        map_no_data: bool,
+        out_no_data: TO,
+    ) -> Result<Vec<TO>>;
+}
+
+#[async_trait]
+impl<TO, T1> ExpressionTupleProcessor<TO> for BoxRasterQueryProcessor<T1>
+where
+    TO: Pixel,
+    T1: Pixel + AsPrimitive<TO>,
+{
+    type Tuple = RasterTile2D<T1>;
+
+    #[inline]
+    async fn queries<'a>(
+        &'a self,
+        query: RasterQueryRectangle,
+        ctx: &'a dyn QueryContext,
+    ) -> Result<BoxStream<'a, Result<Self::Tuple>>> {
+        let stream = self.query(query, ctx).await?;
+
+        Ok(stream.boxed())
+    }
+
+    #[inline]
+    fn all_empty(tuple: &Self::Tuple) -> bool {
+        tuple.grid_array.is_empty()
+    }
+
+    #[inline]
+    fn empty_raster(tuple: &Self::Tuple) -> RasterTile2D<TO> {
+        tuple.clone().convert()
+    }
+
+    #[inline]
+    fn metadata(tuple: &Self::Tuple) -> (TimeInterval, GridIdx2D, GeoTransform, GridShape2D) {
+        let raster = &tuple;
+
+        (
+            raster.time,
+            raster.tile_position,
+            raster.global_geo_transform,
+            raster.grid_shape(),
+        )
+    }
+
+    #[inline]
+    fn compute_expression(
+        raster: Self::Tuple,
+        program: &LinkedExpression,
+        map_no_data: bool,
+        out_no_data: TO,
+    ) -> Result<Vec<TO>> {
+        let expression = unsafe {
+            // we have to "trust" that the function has the signature we expect
+            program.function_3::<f64, bool, f64>()?
+        };
+
+        // cannot be empty at this point
+        let tile = raster.into_materialized_tile();
+
+        let data = tile
+            .grid_array
+            .data
+            .par_iter()
```

Comment:
I think we need to chunk here, otherwise the overhead of the task is too high

Reply:
i guess we have to benchmark this

Reply:
It is not too high, at least it outperforms OpenCL. And also the experiments prior to the implementation showed that this is not a major problem.

However, in a separate task, one could measure this and try to improve the performance.